### PR TITLE
Implement paging

### LINKED ISSUE #17: Implement paging
- https://os.phil-opp.com/paging-introduction/
- https://os.phil-opp.com/paging-implementation/

### PR DESCRIPTION
**Pull request by 0xf4lc0n**
Resolves #17 

The [MemoryManager](https://github.com/0xf4lc0n/cosmos/blob/947bf33eb9d0e00de1d560d75c2a8714d83e92d3/kernel/src/memory.rs#L18C1-L21C2) is useful when it comes to testing or abstracting operations that require both `OffsetPageTable` and `BootInfoFrameAllocator`.

But registering it as singleton hidden behind `Once<Mutex<..>>` may be not very convenient.

In future we will need `MemoryManager` inside `alloc` module or any other module that handles heap allocation.

Saying so there will be no reason to use `MemoryManager` outside of `alloc` module. So we can pass `MemoryManager` directly to that module and provide `Once<Mutex<MemoryManager>>` only for test purposes (by using `#[cfg]` attribute).

Feel free to provide your feedback.

### CODE REVIEW

File: kernel/src/memory.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+    /// Returns an iterator over the usable frames specified in [`MemoryRegions`].
+    fn usable_frames(&self) -> impl Iterator<Item = PhysFrame> {
+        let regions = self.memory_regions.iter();
+        // get usable regions
+        let usable_regions = regions.filter(|r| r.kind == MemoryRegionKind::Usable);
+        // map each region to its address range
+        let addr_ranges = usable_regions.map(|r| r.start..r.end);
+        // transform to an iterator of frame start addresses
+        let frame_addresses = addr_ranges.flat_map(|r| r.step_by(4096));
```
Comment (0xf4lc0n) [inline kernel/src/memory.rs:177]: Dont use magic values.

File: kernel/src/memory.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    unsafe fn init(memory_regions: &'static MemoryRegions) -> Self {
+        BootInfoFrameAllocator {
+            memory_regions,
+            next: 0,
+        }
+    }
+
+    /// Returns an iterator over the usable frames specified in [`MemoryRegions`].
+    fn usable_frames(&self) -> impl Iterator<Item = PhysFrame> {
+        let regions = self.memory_regions.iter();
```
Comment (0xf4lc0n) [inline kernel/src/memory.rs:171]: Chain that shit.

File: kernel/src/memory.rs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+        for page in page_range {
+            // Allocate avalible USABLE memory frame
+            let frame = self
+                .frame_allocator
+                .allocate_frame()
+                .ok_or(MapToError::FrameAllocationFailed)?;
+
+            // Map page to allocated memory frame
+            unsafe {
```
Comment (0xf4lc0n) [inline kernel/src/memory.rs:96]: Explain why is it safe.

**Review by Vie0n — APPROVED** [review]
LGTM

**Pull request merged**